### PR TITLE
Add support for MAP data type for C API

### PR DESCRIPTION
**Pull request by WWW0030**
# Description

Added MAP data type to c_api
Added tests to MAP data type

Fixes # (issue)
https://github.com/kuzudb/kuzu/issues/4148

# Contributor agreement

- [ ] I have read and agree to the [Contributor Agreement](https://github.com/kuzudb/kuzu/blob/master/CLA.md).

### CODE REVIEW

File: src/include/common/types/types.h [hunk 1]
```diff
@@ -231,6 +231,7 @@ enum class PhysicalTypeID : uint8_t {
     ARRAY = 23,
     STRUCT = 24,
     POINTER = 25,
+    MAP = 26,
```
Comment (mewim) [inline src/include/common/types/types.h:234]: `MAP` is not a physical type. You should NOT add a physical type for it.

**Comment by mewim** [comment]
The test cases are not enough. You should also add value tests to make sure that a dict returned as query result can actually be used in C API (i.e., you can read from it).

**Review by mewim — CHANGES_REQUESTED** [review]
See my comments

**Comment by mewim** [comment]
Also, I do not think the implmentation is correct. Reading from `MAP` should be done with `NestedVal`. Please refer to https://github.com/kuzudb/kuzu/blob/fae65c679902ea73721784378e64b13421ff237a/tools/python_api/src_cpp/py_query_result.cpp#L280-L289

**Comment by WWW0030** [comment]
> Also, I do not think the implmentation is correct. Reading from `MAP` should be done with `NestedVal`. Please refer to
> 
> https://github.com/kuzudb/kuzu/blob/fae65c679902ea73721784378e64b13421ff237a/tools/python_api/src_cpp/py_query_result.cpp#L280-L289

I think that it is already done with NestedVal inside of the called function `kuzu_value_get_list_element`. The added test seems to be working fine aswell. Please let me know if its OK.

**Comment by mewim** [comment]
> > Also, I do not think the implmentation is correct. Reading from `MAP` should be done with `NestedVal`. Please refer to
> > https://github.com/kuzudb/kuzu/blob/fae65c679902ea73721784378e64b13421ff237a/tools/python_api/src_cpp/py_query_result.cpp#L280-L289
> 
> I think that it is already done with NestedVal inside of the called function `kuzu_value_get_list_element`. The added test seems to be working fine aswell. Please let me know if its OK.

I don't think you called `kuzu_value_get_map_field_name` in the test case. My expected behavior is that `kuzu_value_get_map_field_name` returns `audience1` and `kuzu_value_get_map_field_value` returns a value of `33`. The user SHOULD NOT need to call another `kuzu_value_get_struct_field_name` or `kuzu_value_get_struct_field_value` to read the key or value from the map.

**Review by mewim — CHANGES_REQUESTED** [review]
The ergonomics of the API is not desirable. Further changes are needed.

File: test/c_api/value_test.cpp [hunk 2]
```diff
@@ -556,6 +555,93 @@ TEST_F(CApiValueTest, GetStructFieldValue) {
...
+    ASSERT_EQ(state, KuzuSuccess);
+    ASSERT_TRUE(kuzu_query_result_is_success(&result));
+    ASSERT_TRUE(kuzu_query_result_has_next(&result));
+    state = kuzu_query_result_get_next(&result, &flatTuple);
+    ASSERT_EQ(state, KuzuSuccess);
+    ASSERT_FALSE(kuzu_query_result_has_next(&result));
+    kuzu_value value;
+    ASSERT_EQ(kuzu_flat_tuple_get_value(&flatTuple, 0, &value), KuzuSuccess);
+
+    char* mapName;
```
Comment (mewim) [inline test/c_api/value_test.cpp:599]: There is probably a memory leak here. The `mapName` is not destroyed.

Comment (WWW0030) [reply test/c_api/value_test.cpp:599]: Should be destroyed below, line 602

**Review by mewim — APPROVED** [review]
Can be merged if all CI passes

**Pull request merged**